packages/langium/src/lsp/call-hierarchy-provider.ts

```diff
@@ -0,0 +1,107 @@
+
+    incomingCalls(params: CallHierarchyIncomingCallsParams,  cancelToken?: CancellationToken): CallHierarchyIncomingCall[] | null;
+
+    outgoingCalls(params: CallHierarchyOutgoingCallsParams, cancelToken?: CancellationToken): CallHierarchyOutgoingCall[] | null;
+}
+
+export class DefaultCallHierarchyProvider implements CallHierarchyProvider {
+    protected readonly grammarConfig: GrammarConfig;
+    protected readonly nameProvider: NameProvider;
+    protected readonly documents: LangiumDocuments;
+    protected readonly references: References;
+
+    constructor(services: LangiumServices) {
+        this.grammarConfig = services.parser.GrammarConfig;
+        this.nameProvider = services.references.NameProvider;
+        this.documents = services.shared.workspace.LangiumDocuments;
+        this.references = services.references.References;
+    }
+
+    prepareCallHierarchy(document: LangiumDocument<AstNode>, params: CallHierarchyPrepareParams): CallHierarchyItem[] | null {
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.position), this.grammarConfig.nameRegexp);
+        if (!targetNode) {
+            return null;
+        }
+
+        const declarationNode = this.references.findDeclaration(targetNode);
+        if (!declarationNode) {
+            return null;
+        }
+
+        return this.getCallHierarchyItems(declarationNode?.element, document);
+    }
+
+    protected getCallHierarchyItems(targetNode: AstNode, document: LangiumDocument<AstNode>): CallHierarchyItem[] {
+        const nameNode = this.nameProvider.getNameNode(targetNode);
+        if (!nameNode) {
+            return [];
+        }
+
+        return [{
+            kind: SymbolKind.Method,
+            name: nameNode.text,
+            range: nameNode.range,
+            selectionRange: nameNode.range,
+            uri: document.uri.toString()
+        }];
+    }
+
+    incomingCalls(params: CallHierarchyIncomingCallsParams): CallHierarchyIncomingCall[] | null {
+        const document = this.documents.getOrCreateDocument(URI.parse(params.item.uri));
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.item.range.start), this.grammarConfig.nameRegexp);
+        if (!targetNode) {
+            return null;
+        }
+
+        const references = this.references.findReferences(targetNode.element, {includeDeclaration: false, onlyLocal: false}).toArray();
+        return this.getIncomingCalls(references);
+    }
+
+    /**
+     * Override this method to collect the incoming calls for your language
+     */
+    protected getIncomingCalls(_references: ReferenceDescription[]): CallHierarchyIncomingCall[] | null {
```

Comment:
Suggestion: To create more fine-grained results, the original source node should be passed in here as well.
```suggestion
    protected abstract getIncomingCalls(node: AstNode, references: ReferenceDescription[]): CallHierarchyIncomingCall[] | undefined {
```